internal/controlplane/handlers_entity_instances.go

```diff
@@ -0,0 +1,311 @@
+		return nil, err
+	}
+
+	// Telemetry logging
+	logger.BusinessRecord(ctx).Provider = providerName
+	logger.BusinessRecord(ctx).Project = projectID
+
+	return &pb.GetEntityByNameResponse{
+		Entity: entity,
+	}, nil
+}
+
+// DeleteEntityById deletes an entity instance for a given entity ID
+func (s *Server) DeleteEntityById(
+	ctx context.Context,
+	in *pb.DeleteEntityByIdRequest,
+) (*pb.DeleteEntityByIdResponse, error) {
+	// Parse entity ID
+	entityID, err := uuid.Parse(in.GetId())
+	if err != nil {
+		return nil, util.UserVisibleError(codes.InvalidArgument, "invalid entity ID")
+	}
+
+	projectID := GetProjectID(ctx)
+
+	// Call service to delete entity
+	err = s.entityService.DeleteEntityByID(ctx, entityID, projectID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Telemetry logging
+	logger.BusinessRecord(ctx).Project = projectID
+	logger.BusinessRecord(ctx).Entity = entityID
+
+	return &pb.DeleteEntityByIdResponse{
+		Id: in.GetId(),
+	}, nil
+}
+
+// CreateEntity creates a new entity instance
+func (s *Server) CreateEntity(
+	ctx context.Context,
+	in *pb.CreateEntityRequest,
+) (*pb.CreateEntityResponse, error) {
+	entityCtx := engcontext.EntityFromContext(ctx)
+	projectID := entityCtx.Project.ID
+	providerName := GetProviderName(ctx)
+
+	// Get provider ID from name
+	provider, err := s.providerStore.GetByName(ctx, projectID, providerName)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, util.UserVisibleError(codes.NotFound, "provider not found")
+		}
+		return nil, fmt.Errorf("error getting provider: %w", err)
+	}
+
+	// Get entity type from request
+	entityType := in.GetEntityType()
+	if entityType == pb.Entity_ENTITY_UNSPECIFIED {
+		return nil, util.UserVisibleError(codes.InvalidArgument, "entity type must be specified")
+	}
+
+	// Get name from request
+	name := in.GetName()
+	if name == "" {
+		return nil, util.UserVisibleError(codes.InvalidArgument, "entity name must be specified")
+	}
+
+	// Get properties from request
+	props := in.GetProperties()
+	if props == nil {
+		props = &structpb.Struct{
+			Fields: make(map[string]*structpb.Value),
+		}
+	}
+
+	l := zerolog.Ctx(ctx).With().
+		Str("projectID", projectID.String()).
+		Str("providerID", provider.ID.String()).
+		Str("entityType", entityType.String()).
+		Str("name", name).
+		Logger()
+	ctx = l.WithContext(ctx)
+
+	// Create entity in database
+	tx, err := s.store.BeginTransaction()
+	if err != nil {
+		return nil, fmt.Errorf("error starting transaction: %w", err)
+	}
+	defer func() {
+		if err := tx.Rollback(); err != nil && !errors.Is(err, sql.ErrTxDone) {
+			l.Error().Err(err).Msg("error rolling back transaction")
+		}
+	}()
+
+	qtx := s.store.GetQuerierWithTransaction(tx)
+
+	// Convert pb.Entity to db.Entities
+	dbEntityType, err := entities.EntityTypeToDBType(entityType)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert entity type: %w", err)
+	}
+
+	// Create entity
+	entity, err := qtx.CreateEntity(ctx, db.CreateEntityParams{
```

Comment:
Does this allow creating a repository which Minder doesn't have access to (or doesn't exist)?

I'd suggest keeping `CreateEntity` out of this PR, as I think the create path is substantially more complicated than the "Get" path.

Reply:
You're right, let me remove it to make life easier. Mind if I leave this (or rename to `RegisterEntity` and just throw an error always?

Reply:
I think throwing UNIMPLEMENTED is fine, but I'm not sure how I feel about committing to the parameters of the call when we don't understand the implementation.

(For example, we can use name + type + provider to pick out an entity once it's been registered -- why can't we do that for the registration as well?)

Reply:
I removed the name since it's something that the provider will derive. However, I left entity type since I figured that using the properties to derive that might be fragile. wdyt?